packages/@react-facet/core/src/components/With.tsx

```diff
@@ -0,0 +1,14 @@
+import { ReactElement } from 'react'
+import { useFacetMap, useFacetUnwrap } from '../hooks'
+import { Facet } from '../types'
+
+type WithProps<T> = {
+  data: Facet<T | null | undefined>
+  children: (data: Facet<T>) => ReactElement
+}
+
+export const With = <T,>({ data, children }: WithProps<T>) => {
+  const shouldRenderFacet = useFacetMap((data) => data != null, [], [data])
+  const shouldRender = useFacetUnwrap(shouldRenderFacet)
+  return shouldRender === true ? children(data as Facet<T>) : null
+}
```

Comment:
Not sure if this is any better, but at least it's more explicit
```suggestion
const hasData = <T,>(_: Facet<T | null | undefined>, shouldRender: boolean | NoValue): _ is Facet<T> => {
  return shouldRender === true
}

export const With = <T,>({ data, children }: WithProps<T>) => {
  const shouldRenderFacet = useFacetMap((data) => data != null, [], [data])
  const shouldRender = useFacetUnwrap(shouldRenderFacet)
  return hasData(data, shouldRender) ? children(data) : null
}
```